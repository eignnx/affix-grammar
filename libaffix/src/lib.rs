#[macro_use]
extern crate serde_derive;

pub mod fault;
pub mod gen;
pub mod literate;
pub mod parser;
pub mod semantics;

#[cfg(test)]
mod tests {
    use super::*;
    use std::convert::TryFrom;
    use std::fs;
    use std::path::PathBuf;

    static RNG_SEED: u64 = 123456;
    static MAX_VALUES: usize = 100;

    #[test]
    fn run_snapshots() {
        let inputs = PathBuf::from(std::env!("CARGO_MANIFEST_DIR"))
            .join("src")
            .join("snapshot_inputs");

        for dir_entry in fs::read_dir(inputs).unwrap() {
            let path = dir_entry.unwrap().path();
            let src_text = fs::read_to_string(&path).unwrap();
            let parse_res = parser::syntax::ParsedGrammar::try_from(&src_text[..]);

            let parse_snapshot_name = format!(
                "PARSE_SNAPSHOT__{}",
                path.file_stem().unwrap().to_string_lossy()
            );
            insta::assert_json_snapshot!(parse_snapshot_name, parse_res);

            if let Some(grammar) = parse_res.ok() {
                let generator = gen::Generator::new_seeded(grammar, RNG_SEED);
                let generated = generate_n(generator, MAX_VALUES);
                let gen_snapshot_name = format!(
                    "GEN_SNAPSHOT__{}",
                    path.file_stem().unwrap().to_string_lossy()
                );
                insta::assert_json_snapshot!(gen_snapshot_name, generated);
            }
        }
    }

    fn generate_n<R: rand::Rng>(
        mut gen: gen::Generator<R>,
        n: usize,
    ) -> fault::DynamicRes<Vec<String>> {
        let mut sentences = vec![];
        for _ in 0..n {
            match gen.generate() {
                Ok(sentence) => sentences.push(sentence),
                Err(fault::DynamicErr::MaxTrialsExceeded { .. }) => return Ok(sentences),
                Err(e) => return Err(e),
            }
        }
        Ok(sentences)
    }
}
