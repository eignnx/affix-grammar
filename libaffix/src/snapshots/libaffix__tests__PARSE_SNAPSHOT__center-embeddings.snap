---
source: libaffix/src/lib.rs
expression: parse_res
---
{
  "Ok": {
    "data_decls": [
      {
        "name": "Subj",
        "variants": {
          "dog": [
            [
              {
                "StrLit": "dog"
              }
            ]
          ],
          "cat": [
            [
              {
                "StrLit": "cat"
              }
            ]
          ],
          "pterodactly": [
            [
              {
                "StrLit": "pterodactyl"
              }
            ]
          ],
          "giraffe": [
            [
              {
                "StrLit": "giraffe"
              }
            ]
          ],
          "ostrich": [
            [
              {
                "StrLit": "ostrich"
              }
            ]
          ],
          "porcupine": [
            [
              {
                "StrLit": "porcupine"
              }
            ]
          ],
          "dash-hound": [
            [
              {
                "StrLit": "dash hound"
              }
            ]
          ],
          "aardvark": [
            [
              {
                "StrLit": "aardvark"
              }
            ]
          ],
          "hamster": [
            [
              {
                "StrLit": "hamster"
              }
            ]
          ],
          "sea-turtle": [
            [
              {
                "StrLit": "sea turtle"
              }
            ]
          ],
          "porpoise": [
            [
              {
                "StrLit": "porpoise"
              }
            ]
          ]
        }
      }
    ],
    "rule_decls": [
      {
        "signature": {
          "name": "start",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "RuleRef": {
                    "rule": "center-embedding",
                    "args": []
                  }
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "center-embedding",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": []
                  }
                },
                {
                  "RuleRef": {
                    "rule": "verb-phrase",
                    "args": []
                  }
                },
                "Plus",
                {
                  "StrLit": "."
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "verb-phrase",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "slept"
                }
              ],
              [
                {
                  "StrLit": "killed"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": []
                  }
                }
              ],
              [
                {
                  "StrLit": "longed for the sea"
                }
              ],
              [
                {
                  "StrLit": "was murdered on the high seas"
                }
              ],
              [
                {
                  "StrLit": "visited"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": []
                  }
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "subject",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "RuleRef": {
                    "rule": "article",
                    "args": [
                      {
                        "Variable": [
                          "Subj",
                          ""
                        ]
                      }
                    ]
                  }
                },
                {
                  "DataVariable": [
                    "Subj",
                    ""
                  ]
                }
              ],
              [
                {
                  "RuleRef": {
                    "rule": "article",
                    "args": [
                      {
                        "Variable": [
                          "Subj",
                          ""
                        ]
                      }
                    ]
                  }
                },
                {
                  "DataVariable": [
                    "Subj",
                    ""
                  ]
                },
                {
                  "RuleRef": {
                    "rule": "that",
                    "args": []
                  }
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": []
                  }
                },
                {
                  "RuleRef": {
                    "rule": "intermediate-verb",
                    "args": []
                  }
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "article",
          "parameter_types": [
            "Subj"
          ]
        },
        "cases": [
          {
            "guard": {
              "requirements": [
                {
                  "Variant": "ostrich"
                }
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "an"
                }
              ],
              [
                {
                  "StrLit": "the"
                }
              ]
            ]
          },
          {
            "guard": {
              "requirements": [
                {
                  "Variant": "aardvark"
                }
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "an"
                }
              ],
              [
                {
                  "StrLit": "the"
                }
              ]
            ]
          },
          {
            "guard": {
              "requirements": [
                "Wild"
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "a"
                }
              ],
              [
                {
                  "StrLit": "the"
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "that",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "that"
                }
              ],
              [
                {
                  "StrLit": "who"
                }
              ],
              [
                {
                  "StrLit": ""
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "intermediate-verb",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "loved"
                }
              ],
              [
                {
                  "StrLit": "spat upon"
                }
              ],
              [
                {
                  "StrLit": "saw"
                }
              ],
              [
                {
                  "StrLit": "intended to kill"
                }
              ],
              [
                {
                  "StrLit": "arrested"
                }
              ],
              [
                {
                  "StrLit": "plucked"
                }
              ],
              [
                {
                  "StrLit": "avoided"
                }
              ],
              [
                {
                  "StrLit": "was attracted to"
                }
              ],
              [
                {
                  "StrLit": "had observed carefully"
                }
              ],
              [
                {
                  "StrLit": "annoyed"
                }
              ],
              [
                {
                  "StrLit": "consumed"
                }
              ],
              [
                {
                  "StrLit": "minimized"
                }
              ],
              [
                {
                  "StrLit": "treated harshly in the workplace"
                }
              ],
              [
                {
                  "StrLit": "patronized"
                }
              ],
              [
                {
                  "StrLit": "complemented"
                }
              ],
              [
                {
                  "StrLit": "insulted"
                }
              ]
            ]
          }
        ]
      }
    ]
  }
}
